bech32/bech32.go

```diff
@@ -0,0 +1,211 @@
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package bech32
+
+import (
+	"fmt"
+	"strings"
+)
+
+const charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+var gen = []int{0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3}
+
+// Decode decodes a bech32 encoded string, returning the human-readable
+// part and the data part excluding the checksum.
+func Decode(bech string) (string, []byte, error) {
+	// Only	ASCII characters between 33 and 126 are allowed.
+	for i := 0; i < len(bech); i++ {
+		if bech[i] < 31 || bech[i] > 126 {
+			return "", nil, fmt.Errorf("invalid character in string")
+		}
+	}
+
+	// The characters must be either all lowercase or all uppercase.
+	lower := strings.ToLower(bech)
+	upper := strings.ToUpper(bech)
+	if bech != lower && bech != upper {
+		return "", nil, fmt.Errorf("string not all lowercase or all uppercase")
+	}
+
+	// We'll work with the lowercase string from now on.
+	bech = lower
+
+	// The string is invalid if the last '1' is non-existent, it is the first
+	// character of the string (no human-readable part) or one of the last 6
+	// characters of the string (since checksum cannot contain '1'), or if the
+	// string is more than 90 characters in total.
+	one := strings.LastIndexByte(bech, '1')
+	if one < 1 || one+7 > len(bech) {
+		return "", nil, fmt.Errorf("invalid index of 1")
+	}
+	if len(bech) > 90 {
+		return "", nil, fmt.Errorf("string too long")
+	}
+
+	// The human-readable part is everything before the last '1'
+	hrp := bech[:one]
+	data := bech[one+1:]
+
+	// Each character corresponds to the byte with value of the index in 'charset'
+	var decoded []byte
+	for i := 0; i < len(data); i++ {
+		index := strings.IndexByte(charset, data[i])
+		if index < 0 {
+			return "", nil, fmt.Errorf("invalid character in data part")
+		}
+		decoded = append(decoded, byte(index))
+	}
+
+	if !bech32VerifyChecksum(hrp, decoded) {
+		return "", nil, fmt.Errorf("checksum failed")
+	}
+
+	// We exclude the last 6 bytes, which is the checksum.
+	return hrp, decoded[:len(decoded)-6], nil
+}
+
+// Encode encodes a byte slice into a bech32 string with the
+// human-readable part hrb. Note that the bytes must each encode 5 bits
+// (base32).
+func Encode(hrp string, data []byte) (string, error) {
+	// Calculate the checksum of the data and append it at the end.
+	checksum := bech32Checksum(hrp, data)
+	combined := append(data, checksum...)
+
+	// The resulting bech32 string is the concatenation of the hrp, the separator
+	// 1, data and checksum. Everything after the separator is represented using
+	// the specified charset.
+	result := []byte(hrp)
+	result = append(result, '1')
+	for _, b := range combined {
+		if int(b) >= len(charset) {
+			return "", fmt.Errorf("invalid data byte: %v", b)
+		}
+		result = append(result, charset[b])
+	}
+	return string(result), nil
+}
+
+// ConvertBits converts a byte slice where each byte is encoding fromBits bits,
```

Comment:
This is a nice utility function, but from a user's point of view, I'd expect it to be called from within the `Encode` function. As it now, users _must_ manually call this function, otherwise they won't properly be encoding in the `bech32` format.